### PR TITLE
feat(#2225): add renderer.hidden_display to show a summary of hidden files below the tree

### PR DESCRIPTION
**Pull request by evertonse**
This PR introduces the rendering of hidden count files per folder, a feature that closes https://github.com/nvim-tree/nvim-tree.lua/issues/2225

Option ``renderer.hidden_display``, allows users to customize how hidden files count information are displayed per folder.
It can be either a function or  a string. It defaults to "none".
- "none": Doesn't inform anything about hidden files.
- "simple": Shows how many hidden files are in a folder.
- "all": Shows how many files are hidden and the number of hidden files per reason why they're hidden.

#### Video for Ilustration
The video below show that we can trace the ``reason`` for exclusion. Also, shows that it works with ``live filter`` and takes into account ``indent marks`` and ``indent width``. 

https://github.com/user-attachments/assets/c5e7812e-ec95-4298-b82b-a94221cde96e

If a function is provided to ``renderer.hidden_display``, it receives a table ``hidden_count`` where keys are reasons and values are the count of hidden files for that reason.

```lua
hidden_count = {
  bookmark = <num>,
  buf = <num>,
  custom = <num>,
  dotfile = <num>,
  git = <num>,
  live_filter = <num>,
}
```
If an error occurs inside this function, we issue a warning:

https://github.com/user-attachments/assets/118a3029-630e-4258-9c93-19e7fe51ac3c

Also exposes the highlight "NvimTreeHiddenDisplay".

## Minimal Setup 
Setup using custom function
```lua
require('nvim-tree').setup {
  renderer = {
    -- hidden_display = 'all',
    hidden_display = function(hidden_count)
      local total_count = 0
      for reason, count in pairs(hidden_count) do
        total_count = total_count + count
      end

      if total_count > 0 then
        return '> ' .. tostring(total_count) .. ' le hidden'
      end
      return nil
    end,
  },
}
```

### CODE REVIEW

**Comment by evertonse** [comment]
DRAFT: Because there's an ordering problem when sorting by depth. Should be usable but I'll try and fix this tomorrow and set it to ready for review.

File: lua/nvim-tree/explorer/reload.lua [hunk 1]
```diff
@@ -102,7 +113,8 @@ function M.reload(node, git_status)
     ---@type uv.fs_stat.result|nil
     local stat = vim.loop.fs_stat(abs)
 
-    if not explorer.filters:should_filter(abs, stat, filter_status) then
+    local filter_reason = explorer.filters:should_filter_as_reason(abs, stat, filter_status)
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/reload.lua:116]: We're getting payback already @raaymax  :)

File: doc/nvim-tree-lua.txt [hunk 2]
```diff
@@ -878,6 +879,46 @@ Number of spaces for an each tree nesting level. Minimum 1.
...
+  buf = <num>,
+  custom = <num>,
+  dotfile = <num>,
+  git = <num>,
+  live_filter = <num>,
+}
+
+Example of function that can be passed:
+
+function(hidden_count)
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:909]: I like this.

File: doc/nvim-tree-lua.txt [hunk 3]
```diff
@@ -878,6 +879,46 @@ Number of spaces for an each tree nesting level. Minimum 1.
...
+  local total_count = 0
+  for reason, count in pairs(hidden_count) do
+    total_count = total_count + count
+  end
+
+  if total_count > 0 then
+    return "(" .. tostring(total_count) .. " hidden")"
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:916]:
```suggestion
    return "(" .. tostring(total_count) .. " hidden)"
```

File: lua/nvim-tree/explorer/filters.lua [hunk 4]
```diff
@@ -223,4 +224,33 @@ function Filters:should_filter(path, fs_stat, status)
     or bookmark(self, path, fs_stat and fs_stat.type, status.bookmarks)
 end
 
+--- Check if the given path should be filtered, and provide the reason why it was
+---@param path string Absolute path
+---@param fs_stat uv.fs_stat.result|nil fs_stat of file
+---@param status table from prepare
+---@return FILTER_REASON
+function Filters:should_filter_as_reason(path, fs_stat, status)
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/filters.lua:232]: I reckon `function Filters:should_filter(path, fs_stat, status)` should go away and be completely replaced by this new function.

File: lua/nvim-tree/node.lua [hunk 5]
```diff
@@ -21,6 +21,7 @@
 ---@field group_next Node|nil
 ---@field nodes Node[]
 ---@field open boolean
+---@field hidden_count table -- Each field of this table is a key for source and value for count
```
Comment (alex-courtis) [inline lua/nvim-tree/node.lua:24]: Could we please rename this to `hidden_stats` ?

I can see myself thinking it's a number...

File: lua/nvim-tree/renderer/init.lua [hunk 6]
```diff
@@ -50,6 +50,14 @@ local function _draw(bufnr, lines, hl_args, signs, extmarks)
       })
     end
   end
+
+  for line_nr, vlines in pairs(virtual_lines) do
+    vim.api.nvim_buf_set_extmark(bufnr, namespace_extmarks_id, line_nr, 0, {
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/init.lua:55]: Usual paranoia - should we create another namespace for this?

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This is just fantastic! It works, code looks good from a high level, I trust it's solid.

Small nits and codestyle, check etc. needed.

File: doc/nvim-tree-lua.txt [hunk 7]
```diff
@@ -878,6 +879,46 @@ Number of spaces for an each tree nesting level. Minimum 1.
...
+  end
+  return nil
+end
+
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:920]:
```suggestion
*nvim-tree.renderer.hidden_display
Determines the rendering of hidden files in a folder.
  Type: `function | string`, Default: `"none"`

  Possible string values are:
    - `"none"`:   Doesn't inform anything about hidden files.
    - `"simple"`: Shows how many hidden files are in a folder.
    - `"all"`:    Shows how many files are hidden and the number of hidden
                files per reason why they're hidden.

  Example `"all"`:
    If a folder has 14 hidden items for various reasons, the display might
    show: >
    (14 total git: 5, dotfile: 9)
<
  If a function is provided, it receives a table `hidden_count` where keys are
  reasons and values are the count of hidden files for that reason.

  The `hidden_count` argument is structured as follows, where <num> is the
  number of hidden files related to the field: >
    hidden_count = {
      bookmark = <num>,
      buf = <num>,
      custom = <num>,
      dotfile = <num>,
      git = <num>,
      live_filter = <num>,
    }
<
  Example of function that can be passed: >
    function(hidden_count)
      local total_count = 0
      for reason, count in pairs(hidden_count) do
        total_count = total_count + count
      end

      if total_count > 0 then
        return "(" .. tostring(total_count) .. " hidden")"
      end
      return nil
    end
<
```

Formatting.

**Comment by alex-courtis** [comment]
This opens up the door to an inline hidden count, like the size information. For another day.

File: lua/nvim-tree/node.lua (hunk 5 above)
Comment (evertonse) [reply lua/nvim-tree/node.lua:24]: Will do

File: lua/nvim-tree/renderer/init.lua (hunk 6 above)
Comment (evertonse) [reply lua/nvim-tree/renderer/init.lua:55]: I'll make sure of that this week

**Comment by evertonse** [comment]
> This opens up the door to an inline hidden count, like the size information. For another day.

I thought the same.

File: lua/nvim-tree/explorer/filters.lua (hunk 4 above)
Comment (evertonse) [reply lua/nvim-tree/explorer/filters.lua:232]: We could. But there's still one place that uses the ``should_filter`` function that doesn't require a ``reason``.
It's located at ``search-node.lua`` file.

**Comment by evertonse** [comment]
Fixed code style and added a separate ``namespace`` for **virtual lines**.

File: lua/nvim-tree/renderer/init.lua (hunk 6 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/init.lua]: TY

File: lua/nvim-tree/explorer/filters.lua (hunk 4 above)
Comment (alex-courtis) [reply lua/nvim-tree/explorer/filters.lua:232]: Got it.

**Review by alex-courtis — APPROVED** [review]
Many thanks for your contribution!

**Pull request merged**